src/backend/catalog/index.c

```diff
@@ -2457,16 +2457,22 @@ IndexBuildScan(Relation parentRelation,
 	 * concurrent build, or during bootstrap, we take a regular MVCC snapshot
 	 * and index whatever's live according to that.
 	 *
-	 * If the relation is an append-only table, we also use a regular MVCC
-	 * snapshot.
+	 * If the relation is an append-only table, we use SnapshotSelf. We can't
+	 * use lightweight SnapshotAny, because it will lead to error on inserting
+	 * to block directory. We can't use regular MVCC snapshot, because it was
+	 * taken before relation lock, and doesn't respect transaction, which was
+	 * committed after snapshot, but before lock aquaring.
```

Comment:
> We can't use lightweight SnapshotAny, because it will lead to error on inserting to block directory.

It's worth to point out the error details (what exactly will be corrupted) or conceptually to show unsoundness of SnapshotAny use case (smth like, we have to read only one version of row from auxiliary relations for AO table) or withdraw from mention this at all.

> We can't use regular MVCC snapshot, because it was taken before relation lock

I think here we have to write down about the relation lock - it's ShareLock that forbids concurrent relation modifications. Therefore, any concurrent modifications that were committed after snapshot acquiring but before this lock granting will be lost under index creating.

> but before lock aquaring

typo, acquiring

Reply:
> ... regular MVCC snapshot, because it was taken before relation lock, and doesn't respect transaction, which was committed after snapshot, but before lock aquaring.

The statement about outdated MVCC snapshot is not correct here because the `GetSnapshotData()` call (this is called under `GetTransactionSnapshot()` call) updates `CurrentSnapshot` data that is returned to caller side (in our case `IndexBuildScan()` call) with new data about completed and in-progress local and global xids. As a consequence, using such snapshot haven't lead to described anomaly. And this is so when:
* CREATE INDEX stmt and concurrent inserts are executed locally in utility mode on segment or
* CREATE INDEX is global but inserts are local or
* inserts are global but CREATE INDEX is local.

The problem in the case when CREATE INDEX and inserts are both global is in dispatching of outdated [global snapshot](https://github.com/arenadata/gpdb/blob/6.17.5_arenadata26/src/backend/cdb/dispatcher/cdbdisp_dtx.c#L160-L165) (that [was saved in the top of stack as ActiveSnapshot](https://github.com/arenadata/gpdb/blob/6.17.5_arenadata26/src/backend/tcop/pquery.c#L1371) before ShareLock acquiring) to segments.